README
PyarrFS - a RAR reading file system
===================================

PyarrFS is a FUSE based file system to access the content of RAR archives as if
they were just normal files and directories. PyarrFS owes technically to FUSE
and rarfile, while the idea originates out of YarrFS - a similar file system
written a few years back by a colleague (and based on my idea).

This code was inspired by the xmp.py example written by Jeff Epler and Csaba
Henk. Thank you!

Thanks to the folks who wrote FUSE and thanks to Marko Kreen for putting
together rarfile!


PyarrFS is developed on Ubuntu and is thus known to work on that platform
together with fairly recent versions of Debian. Other Linux platforms are also
likely to work. Have you tried it on a different platform, please do tell :)


NEWS
Version 0.9.0 (2017-12-14)
--------------------------
Only ~four years since the last release :P
 * Add pass through of extended attributes, which will hopefully enable facl (file ACLs) to work.

Version 0.8.1 (2013-01-30)
--------------------------
 * Fix incorrect relative file access. This was a problem on Debian 7.0 and probably other platforms.

Version 0.8.0 (2012-04-02)
--------------------------
 * Support for disabling compressed archives to avoid slowdowns.

Version 0.7.2 (2011-08-07)
--------------------------
 * Implemented workaround for bug in rarfile where it returns a date with seconds > 59. Also notified Marko Kreen (author of rarfile) of the bug.

Version 0.7.1 (2011-06-20)
--------------------------
 * Not much of a PyarrFS change, but our required changes to rarfile is now no longer local but have been pushed and accepted into rarfile 2.3. This means we no longer require the hacked rarfile version but instead depend on the regular rarfile - awesome!

Version 0.7.0 (2011-02-02)
--------------------------
 * Reading of normal files! Previously, PyarrFS was only able to read rar files and would try to handle normal files just like it would a rar file, which would obviously not work. That's now fixed!
 * Relative root. The root from which PyarrFS should work and mount in the mount directory can now be specified.

Version 0.6.0 (2011-01-07)
--------------------------
Command line option parsing implemented somewhat. Logging configuration is
improved by the integration with command line options and the root dir can be
set, although it so far has no effect ;)

Version 0.5.0 (2011-01-07)
--------------------------
First version released, previous to this, pyarrfs was only available via git
The thing is basically working but has a few rough edges which will be worked
out before releasing 1.0


t/test-read.py
#!/usr/bin/python

import unittest
import math
import random
import rarfile
import os, sys
import subprocess

try:
	subprocess.Popen("rar", stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell = True)
except:
	print >> sys.stderr, "You do not have the 'rar' binary, please install!"
	sys.exit(1)



class PyarrCheck(unittest.TestCase):

	def setUp(self):
		self.scriptdir = os.path.realpath(os.path.dirname(sys.argv[0]))
		_, self.scriptname = os.path.split(sys.argv[0])
		self.scriptpath = os.path.normpath(os.path.join(self.scriptdir, self.scriptname))
		pathname, scriptname = os.path.split(sys.argv[0])
		self.testdir = os.path.join(self.scriptdir, 'rartest')
		self.rarmntdir = os.path.join(self.testdir, 'rarmnt')
		self.testfiledir = os.path.join(self.testdir, 'testfiles')
		self.testarchivedir = os.path.join(self.testdir, 'testarchives')
		self.pyarrpath = self.scriptdir + '/../pyarrfs.py'

		self.mkdir(self.testdir)
		self.mkdir(self.rarmntdir)
		self.mkdir(self.testfiledir)
		self.mkdir(self.testarchivedir)

		# make sure mount dir is empty
		os.system('fusermount -q -z -u ' + self.rarmntdir)
		try:
			os.system(self.pyarrpath + ' ' + self.rarmntdir)
		except:
			pass

		filedata = [
				[ 'test1',
					'this is testfile1 bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla\n' ],
				[ 'test2',
					'crap crap crap crap\n' ]
				]
		filedata.append(['test3', self.generate_content(200000)])
		self.files = []
		for entry in filedata:
			self.files.append(entry[0])

		self.create_test_files(filedata)

		self.uncompressed_rar_archive = 'testarchive1.rar'
		self.create_uncompressed_rar_archive('testarchive1.rar', self.files)


	def mkdir(self, path):
		if not os.path.exists(path):
			os.mkdir(path)
			self.assertTrue(os.path.exists(path))

	def create_test_files(self, filedata):
		for entry in filedata:
			filename = entry[0]
			filedata = entry[1]
			f = open(os.path.join(self.testfiledir, filename), 'w')
			f.write(filedata)
			f.close()

	def create_uncompressed_rar_archive(self, rarfile, files):
		os.chdir(self.testarchivedir)
		for file in files:
			filepath = os.path.join(self.testfiledir, file)
			cmd = 'rar a -inul -ep -m0 ' + os.path.join(self.testarchivedir, rarfile) + ' ' + filepath
			os.system(cmd)


	def tearDown(self):
		os.chdir(self.scriptdir)
		os.system('fusermount -q -z -u ' + self.rarmntdir)
		import shutil
#		shutil.rmtree(self.testdir)

	def generate_content(self, size = 0, population = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'):
		result = ''
		for i in xrange(0, size):
			result += str(random.choice(population))
		return result


	def generate_content_code(self, size = 0):
		# round off size to nearest number divisible by 10
		ns = size - ((size + 10) % 10)
		result = ''
		for i in xrange(0, size):
			result += "$%09d" % (i)
		return result




	def test_read_sequential(self):
		"""Read the entire file sequentially from start
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			rawf = open(raw_file, 'r')
			rarf = open(rar_file, 'r')
			self.assertEqual(rarf.read(), rawf.read(), 'mismatch in sequential read')
			rarf.close()
			rawf.close()


	def test_seek_whence0_1(self):
		"""Single seek from start (whence = 0) to random location in file
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			file_size = os.path.getsize(raw_file)
			rarf = open(rar_file, 'r')
			rawf = open(raw_file, 'r')
			for i in xrange(0, 10000):
				byte = random.randrange(0, file_size)
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 1')
			rarf.close()
			rawf.close()


	def test_seek_whence0_2(self):
		"""Two reads, first from start (whence = 0) to random byte in first half of file and second seek() from start (whence = 0) to second half of file
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			file_size = os.path.getsize(raw_file)
			rarf = open(rar_file, 'r')
			rawf = open(raw_file, 'r')
			for i in xrange(0, 10000):
				byte = random.randrange(0, math.floor(file_size/2))
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 2')
				byte = random.randrange(math.floor(file_size/2), file_size)
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 2')
			rarf.close()
			rawf.close()


	def test_seek_whence0_3(self):
		"""Two reads, first from start (whence = 0) to random byte in second half of file and second seek() from start (whence = 0) to first half of file
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			file_size = os.path.getsize(raw_file)
			rarf = open(rar_file, 'r')
			rawf = open(raw_file, 'r')
			for i in xrange(0, 10000):
				byte = random.randrange(math.floor(file_size/2), file_size)
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 3')
				byte = random.randrange(0, math.floor(file_size/2))
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 3')
			rarf.close()
			rawf.close()


	def test_seek_whence2_1(self):
		"""Single seek from end (whence = 2) to random location in file
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			file_size = os.path.getsize(raw_file)
			rarf = open(rar_file, 'r')
			rawf = open(raw_file, 'r')
			for i in xrange(0, 10000):
				byte = random.randrange(0, file_size)
				rawf.seek(-byte, 2)
				rarf.seek(-byte, 2)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 2) test 1')
			rarf.close()
			rawf.close()


if __name__ == '__main__':
	unittest.main()



pyarrfs/pyarrfs.py
#!/usr/bin/env python
# vim: et ts=4 :
#
# PyarrFS - a RAR reading file system
# Copyright (c) 2010-2012 Kristian Larsson <kristian@spritelink.net>
#
# This file is licensed under the X11/MIT license, please see the file COPYING,
# distributed with PyarrFS for more details.
#

import os, sys
import errno
import fcntl
import re
import stat
import logging
import logging.handlers

try:
    import fuse
except:
    print >> sys.stderr, "You do not have the Python module for FUSE lib installed"
    import os.path
    if os.path.exists('/etc/debian_version'):
        print >> sys.stderr, "This appears to be a Debian system in which case you should be able to run:"
        print >> sys.stderr, "HINT: sudo apt-get install python-fuse"
    sys.exit(1)

try:
    import rarfile
except:
    print >> sys.stderr, "You do not have the Python module for the library rarfile installed."
    if os.path.exists('/etc/debian_version'):
        print >> sys.stderr, "This appears to be a Debian system in which case you should be able to run:"
        if not os.path.exists('/usr/bin/pypi-install'):
            print >> sys.stderr, "HINT: sudo apt-get install python-stdeb"
        print >> sys.stderr, "HINT: sudo pypi-install rarfile"
    else:
        print >> sys.stderr, "Please use your distributions package manager or easy_install to get it. Note\nthat you need version 2.3 or later."
        print >> sys.stderr, "HINT: sudo easy_install rarfile"
    sys.exit(1)

try:
    import xattr
except:
    print >> sys.stderr, "You do not have the Python module for the library xattr installed."
    if os.path.exists('/etc/debian_version'):
        print >> sys.stderr, "This appears to be a Debian system in which case you should be able to run:"
        print >> sys.stderr, "HINT: sudo apt-get install python-xattr"
    else:
        print >> sys.stderr, "Please use your distributions package manager or easy_install to get it."
        print >> sys.stderr, "HINT: sudo easy_install xattr"
    sys.exit(1)

rarfile.NEED_COMMENTS = 0

__version__         = '0.9.0'
__author__          = 'Kristian Larsson'
__author_email__    = 'kristian@spritelink.net'
__license__         = 'MIT'
__url__             = 'http://labs.spritelink.net/pyarrfs'


# log settings
log_format = "%(levelname)-8s %(message)s"
logger = logging.getLogger()

fuse.fuse_python_api = (0, 2)

fuse.feature_assert('stateful_files', 'has_init')


def isRarFilePath(path):
    if re.match(r'.*\.rar$', path, re.IGNORECASE):
        return True
    return False


def isRarDirPath(path):
    (a, b) = rarDirSplit(path)
    if a is False or b is False:
        return False
    return True

def rarDirSplit(path):
    m = re.match(r'(.*\.rar)/(.+)', path, re.IGNORECASE)
    if m is not None:
        return m.group(1), m.group(2)
    # FIXME: should raise exception instead?
    return False, False


class Pyarr(fuse.Fuse):
    def __init__(self, *args, **kw):
        logger.info("init!")

        fuse.Fuse.__init__(self, *args, **kw)

        self.no_compressed = False
        self.debug = False
        self.pydebug = False
        self.foreground = False
        self.root = '/'



    def fsinit(self):
        """Called once for initialising things after FUSE itself has been brought up
        """
        os.chdir(self.root)



    def access(self, path, mode):
        """Returns whether a user has access to performing certain operations
        """
        logger.info("access -- " + path)

        # test for write access, PyarrFS is incapable of doing writes, so no axx
        if mode == os.W_OK:
            return -errno.EACCES

        # allow the rest
        # FIXME: do more granular access control, based on RAR file?
        if isRarFilePath(path): # it's a rar file
            # writing is already disallowed and for directories we allow both
            # reading and execution for all users, ie all other modes
            return

        if isRarDirPath(path):  # inside rar archive
            # we allow access to everything (except write, which has already
            # been disallowed) inside rar files
            return

        # files not inside of a rar archive, do passthrough to os.access()
        if not os.access('.' + path, mode):
            return -errno.EACCES



    def getattr(self, path):
        """FS equivalent of stat - returns attributes for object at path
        """
        logger.info("getattr -- " + str(path))
        if isRarFilePath(path): # is a rarfile
            logging.debug("getattr: on rar archive for path " + str(path))

            # if we run with the no_compressed option and files in a rar file
            # are compressed, we just present it as a ordinary directory
            if self.no_compressed:
                rf = rarfile.RarFile('.' + path)
                for n in rf.namelist():
                    inf = rf.getinfo(n)
                    if self.no_compressed and int(chr(inf.compress_type)) > 0:
                        return os.lstat('.' + path)

            original_stat = os.lstat('.' + path)
            fake_stat = fuse.Stat()
            fake_stat.st_mode = stat.S_IFDIR | 0755
            fake_stat.st_ino = 0
            fake_stat.st_dev = 0
            fake_stat.st_rdev = 0
            fake_stat.st_nlink = 2
            fake_stat.st_uid = original_stat.st_uid
            fake_stat.st_gid = original_stat.st_gid
            fake_stat.st_size = 4096
            fake_stat.st_atime = original_stat.st_atime
            fake_stat.st_mtime = original_stat.st_mtime
            fake_stat.st_ctime = original_stat.st_ctime
            logging.debug("getattr: returning fake_stat for " + str(path))
            return fake_stat

        elif isRarDirPath(path):    # is inside a rar file
            logging.debug("getattr: we need to check inside rar archive for path " + str(path))
            (rar_file, rar_path) = rarDirSplit(path)

            original_stat = os.lstat('.' + rar_file)
            rf = rarfile.RarFile('.' + rar_file, 'r', None, None, False)
            try:
                rfi = rf.getinfo(rar_path)
            except:
                # FIXME: add DEBUG log entry
                return -errno.ENOENT

            fake_stat = fuse.Stat()
            fake_stat.st_mode = stat.S_IFREG | 0444
            fake_stat.st_ino = 0
            fake_stat.st_dev = 0
            fake_stat.st_rdev = 0
            fake_stat.st_nlink = 1
            fake_stat.st_uid = original_stat.st_uid
            fake_stat.st_gid = original_stat.st_gid
            fake_stat.st_size = rfi.file_size
            fake_stat.st_blocks = (fake_stat.st_size + 511) / 512
            fake_stat.st_blksize = 4096

            import datetime
            (rft_year, rft_month, rft_day, rft_hour, rft_minute, rft_second) = rfi.date_time
            # fix for broken rar archives
            if rft_second > 59:
                rft_second = 59
            rft_time = datetime.datetime(rft_year, rft_month, rft_day, rft_hour, rft_minute, rft_second)
            fake_stat.st_atime = int(rft_time.strftime('%s'))
            fake_stat.st_mtime = int(rft_time.strftime('%s'))
            fake_stat.st_ctime = int(rft_time.strftime('%s'))
            logger.debug("getattr: returning fake_stat for " + str(rar_path) + " inside rar " + str(rar_file))
            return fake_stat
     
        # normal file outside of any rar file
        logger.debug("getattr: returning normal os.lstat() for path " + str(path))
        return os.lstat('.' + path)



    def getxattr(self, path, name, foo):
        """Get extended attributes, we just try to pass shit through. This is
        rather naive but seems to work for facl (tested by getfacl).
        """
        logger.info("getxattr -- path:{} xattr:{} foo:{}".format(path, name, foo))
        if isRarDirPath(path):    # is inside a rar file
            logging.debug("getxattr: we need to check inside rar archive for path " + str(path))
            (rar_file, rar_path) = rarDirSplit(path)

            logger.debug("getxattr: returning xattr for path " + str(rar_file))
            xa = xattr.xattr('.' + rar_file)
            return xa.get(name)

        # normal file outside of any rar file
        logger.debug("getxattr: returning xattr for path " + str(path))
        xa = xattr.xattr('.' + path)
        return xa.get(name)



    def readdir(self, path, offset):
        """ readdir - return directory listing
        """
        logger.info("readdir -- path: " + str(path) + "  offset: " + str(offset) )
        dirent = [ '.', '..' ]

        if isRarFilePath(path):
            logger.debug("readdir: on rar archive, using rarfile")
            rf = rarfile.RarFile('.' + path, 'r', None, None, False)
            for e in rf.namelist():
                dirent.append(str(e))
        else:
            logger.debug("readdir: normal dir, using os.listdir()")
            # TODO: why do we need to try this? weird way to handle it
            try:
                os.listdir('.' + path)
            except:
                return

            for e in os.listdir('.' + path):
                dirent.append(e)

        for e in dirent:
            yield fuse.Direntry(e)



    def readlink(self, path):
        """ path is a symbolic link and readlink returns where it points too

            Symbolic links are not supported within rar files and so we will
            never find one there, thus this is just a wrapper for the normal os
            call readlink().
        """
        return os.readlink('.' + path)



    def statfs(self):
        """ statfs pass-through function

            statfs is called by df (and similar calls?). We do pass-through to
            underlying filesystem as we don't really change anything.
        """
        return os.statvfs('.')



    class PyarrFile(object):
        """ Class representing a file opened somewhere in a PyarrFS file system

            Returning a class to represent a file will result in stateful
            handling of the file from FUSE. The instantiated object will
            persist over the entire period the file is open and any state held
            within will thus also persist.

            This class is used both for non-rar files as well as rar files and
            thus needs to check what kind of file we're dealing with.
        """
        def __init__(self, path, flags, *mode):
            # Enabling direct_io disables the kernels page cache.
            # Since the content of our RAR files should be pretty stable, we do
            # NOT enable this, ie we allow the kernel to cache all data.
            # This means we get both the .rar file and the file inside the RAR 
            # archive in the block cache which might affect performance 
            # negatively. On the other hand, the kernel could purge the .rar 
            # file from its block cache and just keep the inside file. 
            # It's likely mostly a performance thing. I'm guessing it's good
            # to let the kernel decide on what to cache and what not too, so we
            # set it to False. Some benchmarking should really be done to prove
            # either setting is better than the other
            self.direct_io = False
            # keep_cache means the kernel is allowed to cache content of a file
            # after its been closed an reopened. You only want to disable this
            # if the content of the file might change without the kernel knowing
            # about it, which would be typical for a networked file system.
            # That's not the case with PyarrFS so we enable it.
            self.keep_cache = True

            if isRarDirPath(path):
                (rar_file, rar_path) = rarDirSplit(path)
                self.rf = rarfile.RarFile('.' + rar_file, 'r', None, None, False)
                self.file = self.rf.open(rar_path)
            else:
                self.file = open('.' + path)


        def read(self, length, offset):
            """ read length amount of data from a file and from a given offset
            """
            self.file.seek(offset)
            return self.file.read(length)


        def release(self, flags):
            """ release, or close, a file
            """
            self.file.close()



    def main(self, *a, **kw):
        self.file_class = self.PyarrFile
        return fuse.Fuse.main(self, *a, **kw)




def main():
    usage = """
PyarrFS mirror the filesystem tree from some point on, allowing RAR archives to be treated as directories and files within those RAR archives to be read as regular files.

""" + fuse.Fuse.fusage

    server = Pyarr(version="PyarrFS " + __version__,
                 usage=usage)

    # TODO: what does multithreaded really mean?
    #       just run many requests in parallel? We have no real dependancies
    #       between calls so we could probably enable this to increase
    #       throughput in environments with lots of simultaneous requests.
    #       Looks like it needs PyarrFile to be locked, what's the point then?
    server.multithreaded = False

    server.parser.add_option('-r', '--root', dest='root', metavar="PATH", default=server.root, help="mirror filesystem from under PATH [default: %default]")
    server.parser.add_option('-n', '--no-compressed', action='store_true', dest='no_compressed', default=False, help="Disable compressed files")
    server.parser.add_option('-D', '--pydebug', action='store_true', dest='pydebug', default=False, help="enable debug for just PyarrFS (not FUSE) (implies -f)")
    server.parse(values=server, errex=1)

    # always log to syslog
    if sys.platform == 'darwin':
        log_syslog = logging.handlers.SysLogHandler(address = '/var/run/syslog')
    else:
        log_syslog = logging.handlers.SysLogHandler(address = '/dev/log')

    log_syslog.setFormatter(logging.Formatter(log_format))
    log_syslog.setLevel(logging.WARNING)
    logger.addHandler(log_syslog)

    opts = {}
    # FIXME: why doesn't optdict work?
    for o in server.parser.fuse_args.optlist:
        opts[o] = True

    if server.pydebug:
        server.parser.fuse_args.modifiers['foreground'] = True
        logger.setLevel(logging.DEBUG)

    if server.parser.fuse_args.modifiers['foreground'] or opts.has_key('debug') or server.pydebug:
        log_stream = logging.StreamHandler()
        log_stream.setFormatter(logging.Formatter("%(asctime)s: " + log_format))
        log_stream.setLevel(logging.DEBUG)
        if opts.has_key('debug') or server.pydebug:
            # FIXME: this isn't working, why? we only get WARNING messages
            log_stream.setLevel(logging.DEBUG)
        logger.addHandler(log_stream)

    # and go!
    if server.parser.fuse_args.mountpoint is None:
        print >> sys.stderr, "ERROR: No mount point specified!\n"
        server.parser.print_help()
        sys.exit(1)
    server.main()




if __name__ == '__main__':
    main()




test-read.py
#!/usr/bin/python

import unittest
import math
import random
import rarfile
import os, sys


class PyarrCheck(unittest.TestCase):

	def setUp(self):
		self.scriptdir = os.path.realpath(os.path.dirname(sys.argv[0]))
		_, self.scriptname = os.path.split(sys.argv[0])
		self.scriptpath = os.path.normpath(os.path.join(self.scriptdir, self.scriptname))
		pathname, scriptname = os.path.split(sys.argv[0])
		self.testdir = os.path.join(self.scriptdir, 'rartest')
		self.rarmntdir = os.path.join(self.testdir, 'rarmnt')
		self.testfiledir = os.path.join(self.testdir, 'testfiles')
		self.testarchivedir = os.path.join(self.testdir, 'testarchives')
		self.pyarrpath = '/home/kll/kod/pyarrfs/pyarrfs'

		self.mkdir(self.testdir)
		self.mkdir(self.rarmntdir)
		self.mkdir(self.testfiledir)
		self.mkdir(self.testarchivedir)

		# make sure mount dir is empty
		os.system('fusermount -q -z -u ' + self.rarmntdir)
		try:
			os.system(self.pyarrpath + ' ' + self.rarmntdir)
		except:
			pass

		filedata = [
				[ 'test1',
					'this is testfile1 bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla bla\n' ],
				[ 'test2',
					'crap crap crap crap\n' ]
				]
		filedata.append(['test3', self.generate_content(200000)])
		self.files = []
		for entry in filedata:
			self.files.append(entry[0])

		self.create_test_files(filedata)

		self.uncompressed_rar_archive = 'testarchive1.rar'
		self.create_uncompressed_rar_archive('testarchive1.rar', self.files)


	def mkdir(self, path):
		if not os.path.exists(path):
			os.mkdir(path)
			self.assertTrue(os.path.exists(path))

	def create_test_files(self, filedata):
		for entry in filedata:
			filename = entry[0]
			filedata = entry[1]
			f = open(os.path.join(self.testfiledir, filename), 'w')
			f.write(filedata)
			f.close()

	def create_uncompressed_rar_archive(self, rarfile, files):
		os.chdir(self.testarchivedir)
		for file in files:
			filepath = os.path.join(self.testfiledir, file)
			cmd = 'rar a -inul -ep -m0 ' + os.path.join(self.testarchivedir, rarfile) + ' ' + filepath
			os.system(cmd)


	def tearDown(self):
		os.chdir(self.scriptdir)
		os.system('fusermount -q -z -u ' + self.rarmntdir)
		import shutil
#		shutil.rmtree(self.testdir)

	def generate_content(self, size = 0, population = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz'):
		result = ''
		for i in xrange(0, size):
			result += str(random.choice(population))
		return result


	def generate_content_code(self, size = 0):
		# round off size to nearest number divisible by 10
		ns = size - ((size + 10) % 10)
		result = ''
		for i in xrange(0, size):
			result += "$%09d" % (i)
		return result




	def test_read_sequential(self):
		"""Read the entire file sequentially from start
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			rawf = open(raw_file, 'r')
			rarf = open(rar_file, 'r')
			self.assertEqual(rarf.read(), rawf.read(), 'mismatch in sequential read')
			rarf.close()
			rawf.close()


	def test_seek_whence0_1(self):
		"""Single seek from start (whence = 0) to random location in file
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			file_size = os.path.getsize(raw_file)
			rarf = open(rar_file, 'r')
			rawf = open(raw_file, 'r')
			for i in xrange(0, 10000):
				byte = random.randrange(0, file_size)
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 1')
			rarf.close()
			rawf.close()


	def test_seek_whence0_2(self):
		"""Two reads, first from start (whence = 0) to random byte in first half of file and second seek() from start (whence = 0) to second half of file
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			file_size = os.path.getsize(raw_file)
			rarf = open(rar_file, 'r')
			rawf = open(raw_file, 'r')
			for i in xrange(0, 10000):
				byte = random.randrange(0, math.floor(file_size/2))
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 2')
				byte = random.randrange(math.floor(file_size/2), file_size)
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 2')
			rarf.close()
			rawf.close()


	def test_seek_whence0_3(self):
		"""Two reads, first from start (whence = 0) to random byte in second half of file and second seek() from start (whence = 0) to first half of file
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			file_size = os.path.getsize(raw_file)
			rarf = open(rar_file, 'r')
			rawf = open(raw_file, 'r')
			for i in xrange(0, 10000):
				byte = random.randrange(math.floor(file_size/2), file_size)
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 3')
				byte = random.randrange(0, math.floor(file_size/2))
				rawf.seek(byte)
				rarf.seek(byte)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 0) test 3')
			rarf.close()
			rawf.close()


	def test_seek_whence2_1(self):
		"""Single seek from end (whence = 2) to random location in file
		"""
		for file in self.files:
			rar_file = os.path.normpath(os.path.join(self.rarmntdir, '.' + self.testarchivedir, self.uncompressed_rar_archive, file))
			raw_file = os.path.normpath(os.path.join(self.testfiledir, file))

			file_size = os.path.getsize(raw_file)
			rarf = open(rar_file, 'r')
			rawf = open(raw_file, 'r')
			for i in xrange(0, 10000):
				byte = random.randrange(0, file_size)
				rawf.seek(-byte, 2)
				rarf.seek(-byte, 2)
				self.assertEqual(rarf.read(1), rawf.read(1), 'mismatch in seek (whence = 2) test 1')
			rarf.close()
			rawf.close()


if __name__ == '__main__':
	unittest.main()

